### ISSUE TITLE
 How do I define the number of sources for FastICA decomposition in the MDP.nodes.FastICANode module?

### ISSUE THREAD

**Issue by ymmx2** [mdp-toolkit/mdp-toolkit#91]
I'm trying to set the number of sources for the FastICA decomposition algorithm in the  [MDP][1] module.
In the below example, I have three different sources that I mix with a matrix to get 4 observations. When I apply the FastICA algorithm, it found 4 sources as, by default, the number of sources is taken as equal as the number of observation.

    import matplotlib.pyplot as plt
    import numpy as np
    import scipy as sp
    from mdp.nodes import FastICANode
    
    # time
    Fs=1024
    N=10*Fs
    t = np.arange(N)/Fs
    
    # sources
    s1=np.sin(2*np.pi*0.5*t)
    s2=(np.random.rand(N)-0.5)*0.02
    s3=sp.signal.sawtooth(2 * np.pi * 2 * t, width=0.5)*0.4
    s4=sp.signal.square(2 * np.pi * 5 * t, duty=0.05)
    
    # Mixing matrix
    W =np.random.dirichlet(np.ones((4)),size=(4,3))
    # compute observations
    x = np.dot(W[:,:,1],np.vstack((s1 , s2 , s4)))
    
    #Apply ICA
    ica = FastICANode(g='tanh') 
    ica.execute(x.T)
    Ae = ica.get_recmatrix().T
    S = np.dot(np.linalg.pinv(Ae),x)
    
    #display results
    plt.figure()
    plt.subplot(4,3,1)
    plt.title('True Sources ')
    plt.plot(t,s1)
    plt.subplot(4,3,4)
    plt.plot(t,s2)
    plt.subplot(4,3,7)
    # plt.plot(t,s3)
    # plt.subplot(4,3,10)
    plt.plot(t,s4)
    
    plt.subplot(4,3,2)
    plt.title('Observations ')
    plt.plot(t,x[0,:])
    plt.subplot(4,3,5)
    plt.plot(t,x[1,:])
    plt.subplot(4,3,8)
    plt.plot(t,x[2,:])
    plt.subplot(4,3,11)
    plt.plot(t,x[3,:])
    
    plt.subplot(4,3,3)
    plt.title('ICA results')
    plt.plot(t,S[0,:])
    plt.subplot(4,3,6)
    plt.plot(t,S[1,:])
    plt.subplot(4,3,9)
    plt.plot(t,S[2,:])
    plt.subplot(4,3,12)
    plt.plot(t,S[3,:])
    plt.show()

So at the end I get this:
[![enter image description here][2]][2]

If I tried to set the number of sources with `ica.set_output_dim(3)`:

    
    #Apply ICA
    ica = FastICANode(g='tanh')
    ica.set_output_dim(3)
    ica.execute(x.T) 
     

I get an error saying:

    mdp.signal_node.InconsistentDimException: Output dim are set already (3) (4 given)!

How can I say to the `FastICANode` class how many sources it should find?

  [1]: http://mdp-toolkit.sourceforge.net/api/index.html
  [2]: https://i.stack.imgur.com/VmBa9.png

**Comment by otizonaizit** [comment]
I am afraid that the FastICA algorithm does not allow for finding a pre-determined number of sources. At the lesat in the version implemented in MDP, FastICA always finds as many sources as there are input dimensions...
I think this is the general assumption for linear ICA.

**Comment by ymmx2** [comment]
The FastICA implemented in Matlab and in sklearn.decomposition allow it though.

**Comment by otizonaizit** [comment]
PR to add this functionality to the MDP implementation is very welcome!!! :-)

**Comment by otizonaizit** [comment]
or, you can use the `white_comp` argument, which probably achieves what the implmentation is matlab an in sklearn are doing. From the docs:
```
        :param white_comp: If whitened is False, you can set 'white_comp' to the
            number of whitened components to keep during the
            calculation (i.e., the input dimensions are reduced to
            white_comp by keeping the components of largest variance).
        :type white_comp: int
```

**Comment by ymmx2** [comment]
Yes, that is what I wanted.
setting the white_comp equal to the number of sources I want.
`ica = FastICANode(g='tanh',white_comp=3)`
Thanks a lot.
